### DISCUSSION TITLE
Backup repository

### DISCUSSION THREAD

**Discussion by ghost** [community/community#22472] [New to GitHub]
<p>Hi everyone,</p>
<p>I manage my repositories through the GitHub website and i haven’t found a way to backup my files.<br>
Is there a way to create a automatic backup of my GitHub repositories, maybe like to a local machine?<br>
I want the backup to include the created branches.</p>
<p>Can someone please provide a solution?</p>
<p>Kind regards,</p>

**Comment by jfullstackdev** [comment]
<p>how about this,</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.blog/2018-12-19-download-your-data/">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/47092464/181071293-765be637-6e17-4d98-b869-586f2445b876.png" class="site-icon" width="32" height="32">

      <a href="https://github.blog/2018-12-19-download-your-data/" target="_blank" rel="noopener" title="05:45PM - 19 December 2018">The GitHub Blog – 19 Dec 18</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/361;"><img src="https://user-images.githubusercontent.com/47092464/181071294-af95029a-e077-4f9b-868b-f4bd02121e41.png" class="thumbnail" width="690" height="361" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/7/5/75ee8af40e466bafae8d473c3f648ce62e7dae3d_2_690x361.png, https://aws1.discourse-cdn.com/github/optimized/3X/7/5/75ee8af40e466bafae8d473c3f648ce62e7dae3d_2_1035x541.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/7/5/75ee8af40e466bafae8d473c3f648ce62e7dae3d.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/7/5/75ee8af40e466bafae8d473c3f648ce62e7dae3d_2_10x10.png"></div>

<h3><a href="https://github.blog/2018-12-19-download-your-data/" target="_blank" rel="noopener">Download all of your GitHub data | The GitHub Blog</a></h3>

  <p>Now you can request an archive of your data from your account settings page.</p>

  <p>
    <span class="label1">Est. reading time: 1 minute</span>
  </p>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>go to</p>
<p><a href="https://github.com/settings/admin">https://github.com/settings/admin</a></p>
<p>and start export</p>

**Comment by airtower-luna** [comment]
<p>According to the documentation the “Export account data” is about account metadata, not repositories.</p>
<p>For repositories, the solution is simple: Clone them locally. The clone includes all branches, tags, etc. <em>unless</em> you disable that. Then backup the repository directories like you would any other directories.</p>

**Comment by ghost** [comment]
<p>Yes but how can I automize and schedule it? I don’t want to manually clone the repository.</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>The easiest solution would be to clone manually <em>once</em>, and then set up a cron job (or an equivalent offered by your OS). How to do that depends on your system. Set the job to run <code>git fetch</code> on the GitHub remote in the repositories.</p>
<p>If you want to use separate repositories and not simply backup the ones you’re working with anyway using bare repositories (without a working tree) would make sense.</p>

**Comment by ghost** [comment]
<p>I want the data to store on a local Windows 10 machine.</p>
<p>I am trying to use the Git Bash terminal to pull data from the remote repository, but it isn’t working. There are no new files showing locally. The command I use is “git pull {repository}”</p>
<p>Now I am receiving an error when authenticating with my GitHub account through SSH.</p>
<p>Error: PTY allocation request failed on channel 0<br>
<img src="https://user-images.githubusercontent.com/10137/181071295-63a59825-0cc0-422b-bb96-810618dc1310.jpeg" alt="error" data-base62-sha1="4i0XYAJO6IoFv9IPPSCyBdQ3ANY" width="574" height="112"></p>

**Comment by jfullstackdev** [comment]
<p>you can also download the entire project on GitHub Web, but without the complete history of the git repo, at least you have the backup,</p>
<p>click the Code and you’ll see</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/1/7/179d1b8d76769a7bab85dc3c1e8d794a2a8b1445.png" data-download-href="/uploads/short-url/3mTAaqSW2myPlFq37sb7J17i8M5.png?dl=1" title="image"><img src="https://user-images.githubusercontent.com/47092464/181071296-5de75a84-af73-4660-93d4-0ee0a90b7564.png" alt="image" data-base62-sha1="3mTAaqSW2myPlFq37sb7J17i8M5" width="556" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/1/7/179d1b8d76769a7bab85dc3c1e8d794a2a8b1445_2_556x500.png, https://aws1.discourse-cdn.com/github/optimized/3X/1/7/179d1b8d76769a7bab85dc3c1e8d794a2a8b1445_2_834x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/1/7/179d1b8d76769a7bab85dc3c1e8d794a2a8b1445_2_1112x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/1/7/179d1b8d76769a7bab85dc3c1e8d794a2a8b1445_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1158×1040 157 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>click <code>Download ZIP</code></p>

**Comment by pbiegun** [comment]
<p>Github data can be divided into two groups:</p>
<ul>
<li>code repository (GIT)</li>
<li>metadata (wiki, comments, pull requests, deployment keys, webhooks, labels, milestones, pipelines and projects.</li>
</ul>
<p>Code repository could be backup as simple as running a git clone, as <a class="mention" href="/airtower-luna">@airtower-luna</a> wrote. However, if you want to back up more data, you need to write your code to use Github API, use scripts and actions like <a href="https://github.com/josegonzalez/python-github-backup" class="inline-onebox" rel="noopener nofollow ugc">GitHub - josegonzalez/python-github-backup: backup a github user or organization</a> or dedicated app from the marketplace <a href="https://github.com/marketplace/gitprotect-io/" class="inline-onebox" rel="noopener nofollow ugc">GitProtect.io Backup · GitHub Marketplace · GitHub</a>.</p>
<p>Just choose the solution that suits your needs, and don’t forget to test if you can restore your backups.</p>

**Comment by ghost** [comment]
<p>Yes I want to back-up the code repository and automize it but I am having difficulties like what I have stated in my previous post.</p>

**Comment by pbiegun** [comment]
<aside class="quote no-group" data-username="bawosambo946" data-post="6" data-topic="229881">
<div class="title">
<div class="quote-controls"></div>
 bawosambo946:</div>
<blockquote>
<p>PTY allocation request failed on channel 0</p>
</blockquote>
</aside>
<p>I think it’s the SSH configuration issue those threats should be more helpful:</p>
<ul>
<li><a href="https://superuser.com/questions/407809/ssh-github-cant-connect" rel="noopener nofollow ugc">https://superuser.com/questions/407809/ssh-github-cant-connect</a></li>
<li><a href="https://stackoverflow.com/questions/3844393/what-to-do-about-pty-allocation-request-failed-on-channel-0" rel="noopener nofollow ugc">https://stackoverflow.com/questions/3844393/what-to-do-about-pty-allocation-request-failed-on-channel-0</a></li>
</ul>

**Comment by davidcorbin-atmosera** [comment]
<p>Cloning the Repo does not get all of the items that are not in the git repo per se - Issues, Releases, everything… What is needed is a means to point at any level of GitHub data (Enterprise, Organization, Repo) and say “give me everything”…and then a mechanism to take that file and re-upload it to an equivalent target…</p>
<p><em>In theory</em> this can be written using the API, but multiple attempts (and man weeks of work) have yet to yield a reliable, comprehensive, durable (i.e. will continue to work despite changes GitHub may make) solution… realistically the only wayt to achieve these goals is for it to be internally developed (as open source) by the GuitHub team.</p>

**Comment by pbiegun** [comment]
<p>Handling Github API could be challenging. When you look into details it becomes a full-time project because you need to write the code, maintain and test it. In most cases, it doesn’t make sense to write it on your own as it takes a lot of time which could be used elsewhere.</p>
<p>Github Enterprise has a dedicated feature to backup a cloud version to an on-premise instance - so it’s more like a mirror, not a backup. If you need fast and reliable backup you could check category Backup on Github marketplace with dedicated tools like <a href="http://Gitprotect.io" rel="noopener nofollow ugc">Gitprotect.io</a>: <a href="https://github.com/marketplace/gitprotect-io/" class="inline-onebox" rel="noopener nofollow ugc">GitProtect.io Backup · GitHub Marketplace · GitHub</a>.</p>

**Comment by davidcorbin-atmosera** [comment]
<p><a class="mention" href="/pbiegun">@pbiegun</a>  - That looks very interesting, and I agree with handling this can easily become a full time project (but clearly it is doable since Xopero Software has achieved it :grinning: )</p>
